### ISSUE TITLE
`Boolean` as a type guard

### ISSUE THREAD

**Issue by pelotom** [Microsoft/TypeScript#31164]
## Search Terms

`boolean` `type` `guard`

## Suggestion

A common idiom in JavaScript is to filter an array using `Boolean` as a filter function to remove falsy values. To make this useful in TypeScript, `Boolean` should act as a type guard:

```diff
+type Falsy = false | 0 | '' | null | undefined;

interface BooleanConstructor {
  new (value?: any): Boolean;
- (value?: any): boolean;
+ <A>(value: A | Falsy): x is A;
  readonly prototype: Boolean;
}
```

## Examples

```js
// The type of `numbers` should be `number[]` but currently it is `(number | undefined)[]`.
const numbers = [1, 2, undefined, 3].filter(Boolean);
```

## Checklist

My suggestion meets these guidelines:

* [x] This wouldn't be a breaking change in existing TypeScript/JavaScript code **(I think this is true?)**
* [x] This wouldn't change the runtime behavior of existing JavaScript code
* [x] This could be implemented without emitting different JS based on the types of the expressions
* [x] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, etc.)
* [x] This feature would agree with the rest of [TypeScript's Design Goals](https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals).

**Comment by DanielRosenwasser** [comment]
Related to #24063.

**Comment by DanielRosenwasser** [comment]
And related to #16655

**Comment by jack-williams** [comment]
Think this is implemented by #29955
